Content.Server/Chemistry/EntitySystems/SolutionInjectOnEventSystem.cs

```diff
@@ -0,0 +1,147 @@
+    {
+        TryInjectTargets(injectorEntity, [target], source);
+    }
+
+    /// <summary>
+    /// Filters <paramref name="targets"/> for valid targets and tries to inject a portion of <see cref="BaseSolutionInjectOnEventComponent.Solution"/> into
+    /// each valid target's bloodstream.
+    /// </summary>
+    /// <remarks>
+    /// Targets are invalid if any of the following are true:
+    /// <list type="bullet">
+    ///     <item>The target does not have a bloodstream.</item>
+    ///     <item><see cref="BaseSolutionInjectOnEventComponent.PierceArmor"/> is false and the target is wearing a hardsuit.</item>
+    ///     <item><see cref="BaseSolutionInjectOnEventComponent.BlockSlots"/> is not NONE and the target has an item equipped in any of the specified slots.</item>
+    /// </list>
+    /// </remarks>
+    /// <returns>true if at least one target was successfully injected, otherwise false</returns>
+    private bool TryInjectTargets(Entity<BaseSolutionInjectOnEventComponent> injector, IReadOnlyList<EntityUid> targets, EntityUid? source = null)
+    {
+        // Make sure we have at least one target
+        if (targets.Count == 0)
+            return false;
+
+        // Get the solution to inject
+        if (!_solutionContainer.TryGetSolution(injector.Owner, injector.Comp.Solution, out var injectorSolution))
+            return false;
+
+        // Build a list of bloodstreams to inject into
+        var targetBloodstreams = new List<Entity<BloodstreamComponent>>();
+        foreach (var target in targets)
+        {
+            if (Deleted(target))
+                continue;
+
+            // Yuck, this is way to hardcodey for my tastes
+            // TODO blocking injection with a hardsuit should probably done with a cancellable event or something
+            if (!injector.Comp.PierceArmor && _inventory.TryGetSlotEntity(target, "outerClothing", out var suit) && _tag.HasTag(suit.Value, "Hardsuit"))
+            {
+                // Only show popup to attacker
+                if (source != null)
+                    _popup.PopupEntity(Loc.GetString(injector.Comp.BlockedByHardsuitPopupMessage, ("weapon", injector.Owner), ("target", target)), target, source.Value, PopupType.SmallCaution);
+
+                continue;
+            }
+
+            // Check if the target has anything equipped in a slot that would block injection
+            if (injector.Comp.BlockSlots != SlotFlags.NONE)
+            {
+                var blocked = false;
+                var containerEnumerator = _inventory.GetSlotEnumerator(target, injector.Comp.BlockSlots);
+                while (containerEnumerator.MoveNext(out var container))
+                {
+                    if (container.ContainedEntity != null)
+                    {
+                        blocked = true;
+                        break;
+                    }
+                }
+                if (blocked)
+                    continue;
```

Comment:
The old logic in inject on collide was:
```c#
            if (containerEnumerator.MoveNext(out _)) 
                 return;
```
which didn't work at all? It just checked if the target had any of the slots, but didn't check if anything was actually in them.